compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -394,6 +394,56 @@ public #{declName}(com.microsoft.bond.StructBondType<#{declName}<#{paramList}>>
             else mempty
 
 
+-- builds text for Object.equals(Object) override
+object_equals :: String -> [Field] -> Maybe Type -> Text
+object_equals declName fields structBase = [lt|@Override
+    public boolean equals(Object o) {
+        if (!(o instanceof #{declName})) return false;
+        #{compareBase structBase}
+        final #{declName} other = (#{declName}) o;
+        #{newlineSep 2 compareField fields}
+        return true;
+    }
+|]
+    where
+        compareBase (Just _)    = [lt|if (!(super.equals(o))) return false;|]
+        compareBase Nothing     = mempty
+        compareField Field {..} = [lt|if (!(#{equalsMember fieldType fieldName})) return false;|]
+        equalsMember t f
+            | isJavaPrimitive t = [lt|this.#{f} == other.#{f}|]
+            | otherwise         = [lt|(this.#{f} == null && other.#{f} == null) || (this.#{f} != null && this.#{f}.equals(other.#{f}))|]
+
+
+-- builds text for Object.hashCode() override
+object_hashCode :: [Field] -> Maybe Type -> Text
+object_hashCode fields structBase = [lt|@Override
+    public int hashCode() {
+        int result = 17;
+        #{newlineSep 2 hash hashInputs}
+        return result;
+    }
+|]
+    where
+        hash codeExpr = [lt|result += #{codeExpr};
+        result *= 0xeadbeef;
+        result ^= result >> 16;|]
+
+        hashInputs = (hashBase structBase) ++ hashFields
+
+        hashBase (Just t)  = [[lt|super.hashCode()|]]
```

Comment:
Consider `hashBase = maybe [] (const [[lt|super.hashCode()|]]) structBase`

If you still want to use pattern matching--which I'm OK with--we can drop simplify the `Just` case to `Just _`.